bumpforceclosefee rpc

ziggie1984 opened the pull request:
~~Fixes https://github.com/lightningnetwork/lnd/issues/8837~~

For reviewers please also focus on the following questions I included into the code starting with

```
// Question: 
```

Should we add an Integration Test for this new RPC call? => Yes I will

Thank you in advance 🙏


EDIT: Depends on https://github.com/lightningnetwork/lnd/pull/8946

ziggie1984 commented:
@coderabbitai review

ziggie1984 commented on cmd/lncli/walletrpc_active.go:
```diff
@@ -485,80 +485,47 @@
 	// Validate the channel point.
 	channelPoint := ctx.Args().Get(0)
-	_, err := NewProtoOutPoint(channelPoint)
+	rpcChannelPoint, err := NewProtoOutPoint(channelPoint)
```
The validations happens in the NewProtoOutPoint function, so its already done

ziggie1984 commented:
While writing the itest, I figured out that we would never account for the parent transaction fee rate when doing a Anchor-CPFP transaction. Fixed it in the course of this PR.

alexbosworth commented on lnrpc/walletrpc/walletkit.proto:
```diff
@@ -273,6 +273,14 @@ service WalletKit {
 
+    /* lncli: `wallet bumpforceclosefee`
+   BumpForceCloseFee is an endpoint that allows users to bump the fee of
+   channel force closes. This does only work for channels of type anchor
```
i assume it would work for p2tr channels too

ziggie1984 commented on lntest/harness_miner.go:
```diff
@@ -294,6 +294,7 @@ func (h *HarnessMiner) AssertTxInMempool(txid *chainhash.Hash) *wire.MsgTx {
 
 		for _, memTx := range mempool {
 			// Check the values are equal.
+			msgTx = h.GetRawTransaction(memTx).MsgTx()
```
This is a bugfix, somehow we would not return the transaction.

ziggie1984 commented on itest/lnd_multi-hop_test.go:
```diff
@@ -1936,17 +1950,6 @@ func runMultiHopHtlcAggregation(ht *lntest.HarnessTest,
-	// We will also expect the success transactions, since he learnt the
-	// preimages from Alice. We also expect Carol to sweep her commitment
-	// output.
-	case lnrpc.CommitmentType_LEGACY:
```
Removed, because the `commitWithZeroConf` map does only test channel types which have anchor outputs, so this case would have never been hit.

ziggie1984 commented on itest/lnd_multi-hop_test.go:
```diff
@@ -2068,12 +2078,6 @@ func runMultiHopHtlcAggregation(ht *lntest.HarnessTest,
-	// In case this is a non-anchor channel type, we must mine 2 blocks, as
-	// the nursery waits an extra block before sweeping. Before the blocks
-	// are mined, we should expect to see Bob's commit sweep in the mempool.
-	case lnrpc.CommitmentType_LEGACY:
```
Same as above, this was never tested because we only test channels with anchor outputs.

ziggie1984 commented:
We definitely need to prioritize the flattening of the mulit_hop tests, fixing those takes super long because they include a lot of different special case for different channel types ...

ziggie1984 commented:
waiting for https://github.com/lightningnetwork/lnd/pull/8717, so that itests are not updated twice.

lightninglabs-deploy commented:
@ziggie1984, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@ziggie1984, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@ziggie1984, remember to re-request review from reviewers when ready

yyforyongyu commented:
I think this should be moved to 0.19 since we are adding an RPC?

saubyk commented:
> I think this should be moved to 0.19 since we are adding an RPC?

Need to understand this better. I thought rpc was already added in 0.18 but not doing its function of bumping up the fee

ziggie1984 commented:
This PR needs a final decision before I will proceed fixing the itests, basically the following:

Currently when creating the Force-Close transaction we will not account for the parent fee rate immediately but rely on the  `createRBFCompliantTx` which will at least create a transaction which meets the current mempool conditions. This has in my view several downsides, that's why I think the new approach by accounting the feerate is better:

When force-closing a channel the conf target will never select the right fee-rate which was essentially required but will only reach this feerate several blocks down the road which can be a long time until the desired feerate is reached in case the commitment transaction (parent) is very big.

Moreover when trying to bump the feerate of a closing transaction with this newly introduced RPC we will not be able to specify the exact feerate and therefore this might be a bad user experience.

Based on the decision above when we account for the parent feerate we need to change some itests which are now failing, but if we decide against it those itests remain unchanged. That's why I am waiting for your feedback before proceeding.

Maybe @morehouse and @yyforyongyu and comment on this matter ?

Thank you in advance for your time.

yyforyongyu commented:
> Moreover when trying to bump the feerate of a closing transaction with this newly introduced RPC we will not be able to specify the exact feerate and therefore this might be a bad user experience.

With the new sweeper I don't think it makes sense to specify a feerate anymore. Instead, users should seek to specify a budget, which is more accurate IMO, as in the end you wanna control the fees paid. Previously we achieved so by specifying a feerate, but it was not precise as it's mostly estimation.

ziggie1984 commented:
> I think this should be moved to 0.19 since we are adding an RPC?

Although its a new RPC, the logic was mostly copied from the `lncli` side, and it fixes the bug of bumping force closes with no htlcs, so I think getting this in lnd 18.3 is a good choice :)

ProofOfKeags commented on itest/lnd_multi-hop_test.go:
```diff
@@ -1956,48 +1959,58 @@ func runMultiHopHtlcAggregation(ht *lntest.HarnessTest,
+
+	for _, tx := range minedTx[0].Transactions {
 		txid := tx.TxHash()
 
 		for i := range tx.TxIn {
 			prevOp := tx.TxIn[i].PreviousOutPoint
-			if _, ok := successOuts[prevOp]; ok {
-				successTxs = append(successTxs, &txid)
+			if _, ok := timeoutOuts[prevOp]; ok {
+				timeoutTxs = append(timeoutTxs, &txid)
 
 				break
 			}
+		}
+	}
```
The outer loop here is a `fn.Filter` and the inner one is a `fn.Find`

ziggie1984 replied:
code change is obsolete

ziggie1984 commented:
> This PR needs a final decision before I will proceed fixing the itests, basically the following:
> 
> Currently when creating the Force-Close transaction we will not account for the parent fee rate immediately but rely on the createRBFCompliantTx which will at least create a transaction which meets the current mempool conditions. This has in my view several downsides, that's why I think the new approach by accounting the feerate is better:
> 
> When force-closing a channel the conf target will never select the right fee-rate which was essentially required but will only reach this feerate several blocks down the road which can be a long time until the desired feerate is reached in case the commitment transaction (parent) is very big.
> 
> Moreover when trying to bump the feerate of a closing transaction with this newly introduced RPC we will not be able to specify the exact feerate and therefore this might be a bad user experience.
> 
> Based on the decision above when we account for the parent feerate we need to change some itests which are now failing, but if we decide against it those itests remain unchanged. That's why I am waiting for your feedback before proceeding.

@ProofOfKeags this is the main question I have not sure if you saw it ?

yyforyongyu commented on lnrpc/walletrpc/walletkit_server.go:
```diff
@@ -1122,6 +1127,121 @@ func (w *WalletKit) BumpFee(ctx context.Context,
+		// inputs which can be used to bump a closed unconfirmed
+		// commitment transaction.
+		if sweep.WitnessType != input.CommitmentAnchor &&
+			sweep.WitnessType != input.TaprootAnchorSweepSpend {
```
think there's also the leased channel type that has anchors.

yyforyongyu commented on lnrpc/walletrpc/walletkit_server.go:
```diff
@@ -1122,6 +1127,121 @@ func (w *WalletKit) BumpFee(ctx context.Context,
+
+		// There might be the case when an anchor sweep is confirmed
+		// between fetching the pending sweeps and preparing the sweep
+		// params. We log this case and proceed.
```
👍

yyforyongyu commented on lnrpc/walletrpc/config_active.go:
```diff
@@ -76,4 +78,9 @@ type Config struct {
+
+	// GetWaitingCloseChannel returns the waiting close channel for a
+	// specified channel point.
+	GetWaitingCloseChannel func(chanPoint wire.OutPoint) (
```
hmm think we should avoid passing functions as config values, think `CurrentNumAnchorChans` was a bad start.

ziggie1984 replied:
changed, but did not touch the `CurrentNumAnchorChans ` should probably refactored in a separate PR.

yyforyongyu commented on lnwallet/wallet.go:
```diff
@@ -1157,6 +1157,27 @@ func (l *LightningWallet) CurrentNumAnchorChans() (int, error) {
+
+	// Count all anchor channels that are open or pending
+	// open, or waiting close.
+	chans, err := l.Cfg.Database.FetchWaitingCloseChannels()
```
IMO this method should be moved to `walletkit_server.go`, and we update the `Config` in the wallet server to take a `db *channeldb.ChannelStateDB`. This way we can avoid passing methods as config values.

ziggie1984 replied:
done

yyforyongyu commented on cmd/lncli/walletrpc_active.go:
```diff
@@ -485,99 +483,39 @@ func bumpForceCloseFee(ctx *cli.Context) error {
-	)
+	resp, err := walletClient.BumpForceCloseFee(
+		ctxc, &walletrpc.BumpFeeRequest{
+			Outpoint:    rpcChannelPoint,
```
as mentioned before, think we need a `BumpForceCloseFeeRequest` to make it clear that this outpoint is not the output being swept but the channel point.

ziggie1984 replied:
agree

yyforyongyu commented on docs/release-notes/release-notes-0.18.1.md:
```diff
@@ -26,6 +26,11 @@
+* [Fixed](https://github.com/lightningnetwork/lnd/pull/8843) a case where 
+bumping an anchor channel closing was not possible when no HTLCs were on the
+commitment when the channel was force closed. Moreover, a new walletrpc endpoint
+`BumpForceCloseFee` was introduced in the course of this bugfix.
```
I'd put this part under `RPC Additions`

yyforyongyu commented on itest/lnd_multi-hop_test.go:
```diff
@@ -1921,11 +1924,22 @@ func runMultiHopHtlcAggregation(ht *lntest.HarnessTest,
 
-	// Mine a block to confirm the closing transaction.
-	ht.MineBlocksAndAssertNumTxes(1, 1)
+	// Mine an empty block first to trigger the anchor cpfp transaction of
```
I think we've already tested it in the previous commit so we don't need to test it again?

yyforyongyu reviewed (changes requested):
Thanks for the fix! My main comment is maybe we could break this PR into two instead,
- first PR for fixing the anchor sweep
- second PR for adding the new RPC

As of today, users can already slow down or speed up a force close tx when the anchor input is managed by the sweeper, detailed in [this doc](https://docs.lightning.engineering/lightning-network-tools/lnd/sweeper#hands-on-managing-sweeps). So if we could get part one merged, uses should at least have a way to manage the force closing process.

As for the second part, I think it's a nice addition but it may need more changes, such as a new RPC request.

ziggie1984 commented:
Ok will split into 2 PR's.

Regarding the parent fee rate I am still undecided especially considering the above mentioned flow.

> When force-closing a channel the conf target will never select the right fee-rate which was essentially required but will only reach this feerate several blocks down the road which can be a long time until the desired feerate is reached in case the commitment transaction (parent) is very big.

I will for now remove my addition of the parent weight consideration, so we can move on and merge this PR, however we might run into unpleasant behaviour when our commitment transaction is always broadcasted underestimating the fees.

ziggie1984 commented on contractcourt/channel_arbitrator.go:
```diff
@@ -1323,13 +1323,14 @@ func (c *ChannelArbitrator) sweepAnchors(anchors *lnwallet.AnchorResolutions,
-
-			return nil
+				"sweeping anchor with default deadline",
+				c.cfg.ChanPoint)
```
I think it would not have made a difference but your proposal is more correct.

ziggie1984 replied:
new relevant PR for this comment: https://github.com/lightningnetwork/lnd/pull/8946

ziggie1984 commented on sweep/fee_bumper.go:
```diff
@@ -1247,7 +1247,7 @@ func prepareSweepTx(inputs []input.Input, changePkScript []byte,
 	}
 
-	txFee := estimator.fee()
+	txFee := estimator.feeWithParent()
```
See https://github.com/lightningnetwork/lnd/pull/8843#issuecomment-2247151000, going to build it back but it remains a trade-off.

ziggie1984 commented on itest/lnd_multi-hop_test.go:
```diff
@@ -1805,7 +1805,7 @@ func runMultiHopHtlcAggregation(ht *lntest.HarnessTest,
 		invoiceReq := &invoicesrpc.AddHoldInvoiceRequest{
 			Value:      invoiceAmt,
-			CltvExpiry: thawHeightDelta - 4,
+			CltvExpiry: thawHeightDelta,
```
agree my bad for the bad docs, because we are not considering the parent feerate anymore for the closing sweep transaction, all the itest changes are obsolete.

ziggie1984 commented on itest/lnd_multi-hop_test.go:
```diff
@@ -1921,11 +1924,22 @@ func runMultiHopHtlcAggregation(ht *lntest.HarnessTest,
+	// Trigger the acutal sweeps.
+	ht.MineEmptyBlocks(1)
+
+	// The above sweep will trigger 2 transactions
+	// The timeout transactions of bob.
+	// And the remote spend + Anchor of Carol
+	// The anchor of bob is not swept because its cltv delta is zero.
+	ht.Miner.AssertNumTxsInMempool(2)
```
yes including the parent feerate to the fee bump logic, broke quite a lot of tests, but because I am building it back, no itest changes are required. However I want to point out that these tests are currently very hard to maintain.

ziggie1984 commented on itest/lnd_multi-hop_test.go:
```diff
@@ -1956,48 +1959,58 @@ func runMultiHopHtlcAggregation(ht *lntest.HarnessTest,
+
+	// Since Bob can aggregate the transactions, we expect a single
+	// transaction, that have multiple spends from the commitment.
+	for _, tx := range txes {
+		txid := tx.TxHash()
+
+		for i := range tx.TxIn {
+			prevOp := tx.TxIn[i].PreviousOutPoint
+			if _, ok := successOuts[prevOp]; ok {
+				successTxs = append(successTxs, &txid)
 
 				break
 			}
```
code change is obsolete.

ziggie1984 commented on itest/lnd_multi-hop_test.go:
```diff
@@ -1956,48 +1959,58 @@ func runMultiHopHtlcAggregation(ht *lntest.HarnessTest,
+	// 1. Carol's sweeping tx for her commit and anchor outputs.
+	// 2. Bob's sweeping tx for all timeout HTLCs.
+	// This block will also trigger the htlc-success sweep transaction.
+	minedTx := ht.MineBlocksAndAssertNumTxes(1, 2)
```
code change is obsolete.

ziggie1984 commented:
Addressed all the comments, splitted the PR into two, first 3 commits of this PR belong to the other PR but we need this change so that the new itest will pass. Otherwise removed all the parent fee rate considerations which  had the positive side effect of not breaking all the itests 😄.

ziggie1984 commented:
@coderabbitai review

ProofOfKeags commented on lnrpc/walletrpc/walletkit_server.go:
```diff
@@ -1122,6 +1128,154 @@ func (w *WalletKit) BumpFee(ctx context.Context,
+	// is requested. Depending on the commitment state when force closing
+	// the channel we might have up to 3 commitments to consider when
+	// bumping the fee.
+	commitSet := fn.NewSet(chainhash.Hash{})
```
Why is the 0 hash included by default?

ziggie1984 replied:
great catch: should be `commitSet := fn.NewSet[chainhash.Hash]()`

ProofOfKeags commented on lnrpc/walletrpc/walletkit_server.go:
```diff
@@ -1122,6 +1128,154 @@ func (w *WalletKit) BumpFee(ctx context.Context,
+
+	pendingSweeps := maps.Values(inputsMap)
+
+	// Use a filter function.
```
This comment is unnecessary. If you want to comment it say something like below. The use of the filter function is self-evident, use the comment to guide purpose, rather than method.

```golang
// Discard everything except for the anchor sweeps.
```

ProofOfKeags commented on itest/lnd_multi-hop_test.go:
```diff
@@ -1936,17 +1936,6 @@ func runMultiHopHtlcAggregation(ht *lntest.HarnessTest,
 
 	expectedTxes := 0
 	switch c {
-	// With the closing transaction confirmed, we should expect Bob's HTLC
-	// timeout transactions to be broadcast due to the expiry being reached.
-	// We will also expect the success transactions, since he learnt the
-	// preimages from Alice. We also expect Carol to sweep her commitment
-	// output.
-	case lnrpc.CommitmentType_LEGACY:
```
😈🔪

ProofOfKeags commented on lntest/miner/miner.go:
```diff
@@ -292,15 +293,21 @@ func (h *HarnessMiner) AssertTxInMempool(txid *chainhash.Hash) *wire.MsgTx {
+				txid, mempool)
+		}
+
+		result.WhenSome(func(txid *chainhash.Hash) {
+			msgTx = h.GetRawTransaction(txid).MsgTx()
+		})
```
Is there any version of `wait.NoError` that allows a real return value (other than just the error)? CC @yyforyongyu

ziggie1984 replied:
recreated one see `NoErrorWithValue` lmk what you think.

ziggie1984 replied:
built it back to the original `NoError` func

ProofOfKeags commented on itest/lnd_sweep_test.go:
```diff
@@ -2150,3 +2150,89 @@ func runBumpFee(ht *lntest.HarnessTest, alice *node.HarnessNode) {
+	bumpFeeReq := &walletrpc.BumpForceCloseFeeRequest{
+		ChanPoint:   chanPoint,
+		SatPerVbyte: uint64(newFeeRate.FeePerVByte()),
+		// The budget needs to be high enough to pay for the fee because
+		// the anchor does not have an output value high enough to pay
+		// for itself.
+		Budget: 100000,
```
Any chance we could de-magic this number by formulating it in terms of the fee rates involved?

ziggie1984 replied:
agree good idea.

ProofOfKeags commented on lntest/rpc/wallet_kit.go:
```diff
@@ -254,6 +254,21 @@ func (h *HarnessRPC) BumpFee(
 
+// BumpForceCloseFee makes a RPC call to the node's WalletKitClient and asserts.
+//
+//nolint:lll
```
I actually don't think you need this `nolint` directive anymore afaict.

ziggie1984 replied:
removing the `nolint:lll` gave me an error for the `make lint` cmd tho ?

ProofOfKeags replied:
disregard then. My GH render seemed to indicate otherwise, but nbd.

ProofOfKeags reviewed (commented):
Small comments, but otherwise looks good (contingent on the last PR)

ziggie1984 commented on lnrpc/walletrpc/walletkit_server.go:
```diff
@@ -1122,6 +1128,154 @@ func (w *WalletKit) BumpFee(ctx context.Context,
+
+	// Check whether there was a dangling commitment at the time the channel
+	// was force closed.
+	remoteCommitDiff, err := channel.RemoteCommitChainTip()
+	if err != nil && !errors.Is(err, channeldb.ErrNoPendingCommit) {
+		return nil, err
+	}
```
I don't think that is necessary because we don't expect a pending commit in the normal case.

ziggie1984 commented on cmd/lncli/walletrpc_active.go:
```diff
@@ -506,78 +481,23 @@ func bumpForceCloseFee(ctx *cli.Context) error {
-		fmt.Printf("Bumping fee of %v:%v: %v\n",
-			sweepTxID, sweep.Outpoint.OutputIndex, resp.GetStatus())
-	}
+	fmt.Println(resp.Status)
```
good point

ziggie1984 commented on lnrpc/walletrpc/walletkit_server.go:
```diff
@@ -1122,6 +1128,154 @@
+		return commitSet.Contains(sweep.OutPoint.Hash)
+	}, pendingSweeps)
+
+	// Filter all relevant anchor sweeps and update the sweep request.
+	for _, anchor := range anchors {
+		// Anchor cpfp bump request are predictable because they are
+		// swept separately hence not batched with other sweeps (they
+		// are marked with the exclusive group flag). Bumping the fee
+		// rate does not create any conflicting fee bump conditions.
+		// Either the rbf requirements are met or the bump is rejected
+		// by the mempool rules.
+		params, existing, err := w.prepareSweepParams(
+			&BumpFeeRequest{
+				Outpoint:    lnrpcOutpoint,
+				TargetConf:  in.TargetConf,
+				SatPerVbyte: in.SatPerVbyte,
+				Immediate:   in.Immediate,
+				Budget:      in.Budget,
+			}, anchor.OutPoint, currentHeight,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		// There might be the case when an anchor sweep is confirmed
+		// between fetching the pending sweeps and preparing the sweep
+		// params. We log this case and proceed.
+		if !existing {
+			log.Errorf("Sweep anchor input(%v) not known to the " +
+				"sweeper subsystem")
+			continue
+		}
+
+		_, err = w.cfg.Sweeper.UpdateParams(anchor.OutPoint, params)
+		if err != nil {
+			return nil, err
+		}
+	}
```
no need, because the update of the sweep parameters does not need to be atomic

yyforyongyu reviewed (changes requested):
Very nice🦾

ziggie1984 commented on cmd/lncli/walletrpc_active.go:
```diff
@@ -424,51 +422,40 @@ var bumpForceCloseFeeCommand = cli.Command{
-	Whether this input will be swept immediately. When set to true, the
-	sweeper will sweep this input without waiting for the next batch.`,
+	Whether this cpfp transaction will be triggered immediately. When set to
+	true, the sweeper will consider all currently registered sweeps and
```
I wanted to outline that by setting this to immediate we will not just sweep this force-close anchor output but also all other pending sweeps which are registered with the sweeper but haven't been swept yet.

ziggie1984 replied:
Added this so outline I am referring to pending sweeps: `consider all currently pending registered sweeps`

yyforyongyu replied:
Cool gotya, now that you mention it, maybe we should change the behavior to sweep only the requested input (set) immediately - but that's for future PRs I guess.

ziggie1984 commented on lntest/wait/wait.go:
```diff
@@ -71,6 +71,38 @@ func NoError(f func() error, timeout time.Duration) error {
 
+// NoErrorWithValue runs the function f and returns its value if it succeeds
+// within the timeout
+func NoErrorWithValue[T any](f func() (T, error), timeout time.Duration) (
```
open for both ways, lets see what @ProofOfKeags thinks.

yyforyongyu replied:
Unless we have a strong argument for adding this - there was a plan to remove `Predicate` so the only exported method is `wait.NoError`, which gives a better design as when writing tests we don't need to think twice about which method to use.

ziggie1984 replied:
ok makes sense, built back.

ProofOfKeags replied:
My argument for something like this was just that `wait.NoError` requires the closure passed in to have side effects if we want to get any values out of it. How does `wait.NoError` pass a value out?

ProofOfKeags replied:
Just looked and it looks like it does not. If we want to consolidate it to a single method, we should extend `wait.NoError` to return a value in addition to the error. If we don't want to use the value we can always monomorphize it to `fn.Unit` / `struct {}`

ziggie1984 commented on lntest/miner/miner.go:
```diff
@@ -281,30 +282,28 @@ func (h *HarnessMiner) GetRawTransactionVerbose(
-	var msgTx *wire.MsgTx
-
-	err := wait.NoError(func() error {
+	txid, err := wait.NoErrorWithValue(func() (*chainhash.Hash, error) {
```
it feels not super clean with the `fn.Option` in place, we would first need to check for `IsNone()` and also after for `WhenSome(func ...)` which feels not really elegant to do, with the new method we just call `result.UnwrapOrErr`. That's my argument to introduce the new func, but as said before no strong options.

yyforyongyu replied:
A second look is I don't think we need to declare the `txid`, we just check it's found in `mempool` and return a `nil` in that case, then do `msgTx := h.GetRawTransaction(txid).MsgTx()` should be enough.

ziggie1984 replied:
sgtm

ProofOfKeags replied:
I'm alright with this as an approach here.

ziggie1984 commented on itest/lnd_sweep_test.go:
```diff
@@ -2150,3 +2152,114 @@ func runBumpFee(ht *lntest.HarnessTest, alice *node.HarnessNode) {
+	// crafting the initial cpfp transaction even less will be used but
+	// technically this is the minimum budget we need to bump the fee rate
+	// to the new starting fee rate of the fee function.
+	minBudget := newFeeRate.FeeForWeight(parentWeight+childWeight) -
```
hmm, I am more in favour of the current approach, let's say we change the max_anchor_default value we might run in to a case where the budget is not enough anymore, however no strong feeling about it, @ProofOfKeags wdyt ?

yyforyongyu replied:
I think we can just use a large budget as the "budget is not enough" part has already been tested elsewhere.

ziggie1984 replied:
sgtm

ziggie1984 replied:
sgtm

ProofOfKeags replied:
My previous issue was really around the magic of the number chosen. The code is gone now but the comment read:

```golang
// Choose a budget big enough to handle the fee bump
```

or something to that effect. I didn't understand why it was considered "big enough" and why the number chosen was convincing, which is why I commented on this before.

ziggie1984 replied:
see my comment here, for this magic number: https://github.com/lightningnetwork/lnd/pull/8843#discussion_r1697559365

ziggie1984 commented on itest/lnd_sweep_test.go:
```diff
@@ -2150,3 +2152,114 @@ func runBumpFee(ht *lntest.HarnessTest, alice *node.HarnessNode) {
+	// transaction was triggered. The sweeper increases the fee rate
+	// periodically with every new incoming block and the selected fee
+	// function.
+	ht.AssertNumTxsInMempool(2)
```
Not sure what you mean, but I calculated the weight of a child because there is still not child transaction in the mempool before bumping the close transaction ?

yyforyongyu replied:
yeah if we use an arbitrarily large enough budget then we can make the test setup a bit easier, then we don't need to care about weights here since we are not asserting an expected feerate was used during the process?

ziggie1984 replied:
just chose a large enough budget so we don't need to care about this.

yyforyongyu commented on itest/lnd_sweep_test.go:
```diff
@@ -2150,3 +2152,114 @@ func runBumpFee(ht *lntest.HarnessTest, alice *node.HarnessNode) {
+	// function.
+	ht.AssertNumTxsInMempool(2)
+
+	// Mine both transactions, the closing tx and the anchor cpfp tx.
```
Quite the opposite, that's why I mentioned we don't "need to check the feerate" or "check the specifics" here, we just check the sweeping attempt can be made.

I think it's important to understand what we are testing here. We are not re-testing the fee bumping logic again, either it's the feerate being increased or skipped. Instead, we are testing the new RPC endpoint, that a user can successfully use to bump the force close tx. There are many edge cases we could test, but I think the common one is,
1. force close happened
2. the user wants to fee bump it, but with a small budget, so calls the RPC.
3. the tx is not confirmed in the next block, the user wants to fee bump it again, so calls the RPC again.

We may also need to check that two versions of the anchor sweeping requests are made every time we call this RPC. And assert the sweeping tx is indeed using one of the anchors, not both - to ensure this new RPC won't prematurely remove the exclusive group.

As for neutrino, I think either we made a dedicated test for it (to avoid nested if-else) or we could skip it. I'm leaning towards skipping it.

ziggie1984 replied:
I reworked the itest, now instead of mining the mempool transaction immediately I make sure I call the `BumpForceCloseFee` rpc two times and make sure that both of the bumps use the same anchor input.

Did exclude the neutrino case.

What I am not really sure about is the following:

```
We may also need to check that two versions of the anchor sweeping requests are made every time we call this RPC. And assert the sweeping tx is indeed using one of the anchors, not both - to ensure this new RPC won't prematurely remove the exclusive group.
```

How can I check that actually two sweeps are attempted (local, remote) every time I bump the close fee ? Maybe you have an idea ?

yyforyongyu commented on itest/lnd_sweep_test.go:
```diff
@@ -2150,3 +2152,114 @@ func runBumpFee(ht *lntest.HarnessTest, alice *node.HarnessNode) {
+func testBumpForceCloseFee(ht *lntest.HarnessTest) {
+	// Skip this test for neutrino, as it's not aware of mempool
+	// transactions.
+	if ht.IsNeutrinoBackend() {
```
Just to be clear I was just curious tho.

ProofOfKeags replied:
> ok also running for neutrino now.

I don't think so?

ziggie1984 replied:
yes built it back, because the itest change quite a bit, we are now relying on the mempool in the itest.

ziggie1984 commented on itest/lnd_sweep_test.go:
```diff
@@ -2146,3 +2154,154 @@ func runBumpFee(ht *lntest.HarnessTest, alice *node.HarnessNode) {
+	// 250 sat/kw) => hence we are fullfilling BIP 125 Rule 4, which is
+	// exactly what we want here to test the subsequent calling of the
+	// bumpclosefee rpc.
+	cpfpBudget := 100_000
```
@ProofOfKeags its not super trivial to come up with the right budget value, I tried to explain it in more detail, wdyt ?

ProofOfKeags replied:
non-blocking. If it needs to be magic, make it magic.

ProofOfKeags commented on cmd/lncli/walletrpc_active.go:
```diff
@@ -485,97 +482,49 @@ func bumpForceCloseFee(ctx *cli.Context) error {
-				Immediate:   ctx.Bool("immediate"),
-				SatPerVbyte: ctx.Uint64("sat_per_vbyte"),
-			})
-		if err != nil {
-			return err
-		}
+	case ctx.Bool("immediate"):
+		immediate = true
 
-		// Bump fee of the anchor sweep.
-		fmt.Printf("Bumping fee of %v:%v: %v\n",
-			sweepTxID, sweep.Outpoint.OutputIndex, resp.GetStatus())
+	case ctx.Bool("force"):
+		immediate = true
 	}
```
```golang
if ctx.IsSet("immediate") && ctx.IsSet("force") {
        return fmt.Errorf("cannot set immediate and force flag at " +
               "the same time")
}
immediate := ctx.Bool("immediate") || ctx.Bool("force")
```

ProofOfKeags commented on sweep/sweeper.go:
```diff
@@ -645,6 +645,12 @@ func (s *UtxoSweeper) collector(blockEpochs <-chan *chainntnfs.BlockEpoch) {
 			// If this input is forced, we perform an sweep
 			// immediately.
+			//
+			// TODO(ziggie): Make sure when `immediate` is selected
+			// as a parameter that we only trigger the sweeping of
+			// this specific input rather than triggering the sweeps
+			// of all current pending inputs registered with the
+			// sweeper.
```
👀 Is this an *eventual* TODO? or a TODO for *this PR*?

ziggie1984 replied:
This TODO is planned for a follow up PR, just marked it in the code base so we have it on the radar.

ProofOfKeags commented on lntest/miner/miner.go:
```diff
@@ -292,19 +291,23 @@ func (h *HarnessMiner) AssertTxInMempool(txid *chainhash.Hash) *wire.MsgTx {
 
 	require.NoError(h, err, "timeout checking mempool")
 
+	msgTx := h.GetRawTransaction(txid).MsgTx()
+
 	return msgTx
```
Consolidate this:

```golang
return h.GetRawTransaction(txid).MsgTx()
```

ziggie1984 replied:
totally, will push the change until I receive the final comments from yong.

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@ziggie1984, remember to re-request review from reviewers when ready

yyforyongyu commented on lntest/wait/wait.go:
```diff
@@ -71,6 +71,40 @@ func NoError(f func() error, timeout time.Duration) error {
 	return nil
 }
 
+// NoErrorWithValue runs the function f and returns its value if it succeeds
```
Looks like this can be removed now.

yyforyongyu commented on itest/lnd_sweep_test.go:
```diff
@@ -2154,3 +2154,151 @@ func runBumpFee(ht *lntest.HarnessTest, alice *node.HarnessNode) {
+	chanPoint := chanPoints[0]
+	alice := nodes[0]
+
+	// We need to fund alice with 2 wallet inputs so that we can test to
```
This does remind me something - that we need to check how many UTXOs are there and return an error if it's below 2, otherwise we can only bump one of the anchors below, and the error will only be shown in the logs, which can be a bad UX. Not in this PR tho as we will soon merge it.

ziggie1984 replied:
Ahh ok so you mean we would need 3 wallet inputs in the above case, so that we can be sure that we have enough wallet inputs to at least try to publish the anchor transaction.

Hmm however I think one input for all the anchor sweeps is enough, because as soon as one succeeds we can be sure the others won't make it into the mempool anyways ?

I added a `TODO` to investigate it.

yyforyongyu replied:
For full nodes yes, for neutrino nope - for instance when a local force close happens, and we only have one wallet utxo, that utxo could be used to bump the remote anchor, while an error won't be returned there.

ziggie1984 replied:
Makes sense, you mean we should generalize this behavior because this test for example skips the neutrino case, however you are right if we are not generalizing it we might run into this problem when all of a sudden we want to support the neutrino case here.

yyforyongyu commented on itest/lnd_sweep_test.go:
```diff
@@ -2154,3 +2154,151 @@ func runBumpFee(ht *lntest.HarnessTest, alice *node.HarnessNode) {
+	// by a single sat hence NOT rbfing the anchor sweep every time a new
+	// block is found and a new sweep broadcast is triggered.
+	//
+	// NOTE:
```
nice docs🙏

yyforyongyu commented on docs/release-notes/release-notes-0.18.3.md:
```diff
@@ -54,6 +54,10 @@ commitment when the channel was force closed.
   send payment stream context, or automatically at the end of the timeout period 
   if the user provided `timeout_seconds`.
 
+* [Add a new rpc endpoint](https://github.com/lightningnetwork/lnd/pull/8843)
```
This now needs to be moved to 0.19.0

yyforyongyu reviewed (commented):
Cool just need to remove the unused method and should be good to go.  Will hold on merging it before 0.18.3 is cut later today.

yyforyongyu commented on docs/release-notes/release-notes-0.19.0.md:
```diff
@@ -0,0 +1,56 @@
+# Release Notes
```
🙏

yyforyongyu reviewed (commented):
Looking good, will hold the merging until 0.18.3 is out so we don't need to cherry-pick on the master.

guggero commented:
0.18.3 is on a side branch, there is no merge freeze in effect.
